carla/recourse_methods/catalog/face/library/face_method.py

```diff
@@ -0,0 +1,282 @@
+    :param step: float > 0; step_size for growing spheres
+    :param mode: str; either 'knn' or 'epsilon'
+    :param model: classification model (either tf keras, pytorch or sklearn)
+    :param p_norm: float=>1; denotes the norm (classical: 1 or 2)
+    :param frac: float 0 < number =< 1; fraction of data for which we compute the graph; if frac = 1, and data set large, then compute long
+    :param keys_immutable: list; list of input names that may not be searched over
+    :param radius: float > 0; parameter for epsilon density graph
+    :return: candidate_counterfactual_star: np array (min. cost counterfactual explanation)
+    """
+    # Choose a subset of data for computational efficiency
+    data = choose_random_subset(data, frac, index)
+
+    # ADD CONSTRAINTS by immutable inputs into adjacency matrix
+    # if element in adjacency matrix 0, then it cannot be reached
+    # this ensures that paths only take same sex / same race / ... etc. routes
+    # TODO: CANNOT deal with continuous constraints (e.g. age) yet
+    # TODO: However, authors also do not mention how to go about it
+    for i in range(len(keys_immutable)):
+        immutable_constraint_matrix1, immutable_constraint_matrix2 = build_constraints(
+            data, i, keys_immutable
+        )
+
+    # POSITIVE PREDICTIONS
+    y_predicted = model.predict_proba(data.values)
+    y_predicted = np.argmax(y_predicted, axis=1)
+    y_positive_indeces = np.where(y_predicted == 1)
+
+    # TODO: replace counterfactual search by function, which takes 'mode' as an argument
+    candidate_counterfactuals_star = []
+    if mode == "knn":
+        boundary = 3  # chosen in ad-hoc fashion
+        median = n_neighbors
+        is_knn = True
+
+    elif mode == "epsilon":
+        boundary = 0.10  # chosen in ad-hoc fashion
+        median = radius
+        is_knn = False
+    else:
+        raise ValueError("Only possible values for mode are knn and epsilon")
+
+    neighbors_list = [
+        median - boundary,
+        median,
+        median + boundary,
+    ]
+
+    # obtain candidate targets (CT); two conditions need to be met:
+    # (1) CT needs to be predicted positively & (2) CT needs to have certain "density"
+    # for knn I interpret 'certain density' as sufficient number of neighbours
+    for n in neighbors_list:
+        find_cf(
+            candidate_counterfactuals_star,
+            data,
+            immutable_constraint_matrix1,
+            immutable_constraint_matrix2,
+            index,
+            n,
+            y_positive_indeces,
+            is_knn=is_knn,
+        )
+
+    candidate_counterfactual_star = np.array(candidate_counterfactuals_star)
+
+    # STEP 4 -- COMPUTE DISTANCES between x^F and candidate x^CF; else return NaN
+    if candidate_counterfactual_star.size == 0:
+        candidate_counterfactual_star = np.empty(
+            data.values.shape[1],
+        )
+        candidate_counterfactual_star[:] = np.nan
+
+    else:
```

Comment:
You could avoid the nested if/else conditions here if you return early when no candidates have been found. Then this `else` could be removed and the below conditions de-intended